### PR TITLE
[WIP] Fixing score

### PR DESCRIPTION
**Pull request by kegl**

### CODE REVIEW

File: workflow/scores/precision_recall.py [hunk 1]
```diff
@@ -226,12 +226,11 @@ def __init__(self, name='precision', precision=2, conf_threshold=0.5):
         self.conf_threshold = conf_threshold
 
     def __call__(self, y_true, y_pred, conf_threshold=None):
-        if conf_threshold is None:
-            conf_threshold = self.conf_threshold
```
Comment (jorisvandenbossche) [inline workflow/scores/precision_recall.py]: this should not be removed (you also did not do it for the other occurrences) 

(it's maybe not the nicest code, but when removing it should be solved in another way)

**Comment by kegl** [comment]
Once this change is accepted, all submissions will have to modify their return type (done for full, see https://github.com/ramp-kits/mars_craters/blob/fixing_score/submissions/full/object_detector.py)

**Comment by kegl** [comment]
I added the optional minipatch to scp. I also made the shape an explicit initial parameter, there is no reason for making (224, 224) default.
@aboucaud: please implement minipatch in ospa and precision/recall. If I remember correctly, we decided to make an asymmetric precision/recall: in recall we would only penalize the undetected real craters if their center is in the minipatch, and symmetrically, in precision we would only penalize the pairless detections (detections which correspond to no real crater) if its center is in the minipatch.
OSPA could also be modified accordingly.

**Comment by jorisvandenbossche** [comment]
I understand the rationale (to be more general in the kind of region definition), but I am not sure I would do it for the current RAMP. 
Main reason is because I *think* it will be more work for the participant to do (although maybe minor, but without clear benefit from participant view). Eg now you can also return an array (which will be typically the output of a model) instead of list of tuples. In the model I was working on (the ssd), I will have as output for each image a 2D array (n_pred, 4), so having to convert it to a nested tuple will be an extra step. 
It also makes it in general even more nested ("list of list of tuple of (tuple, conf))"). 

I would just decide from a user point of view. Because if we think the current (x, y, r, c) tuple is easier, there are other ways to deal with this (being robust to different bbox types) under the hood. Eg we could add a keyword to the score types that specifies the bbox spec format, and depending on this the y_pred values are passed to the actual score function. Then this can be specified when initializing the ScoreType inside problem.py.

**Comment by kegl** [comment]
The problem is that as soon as we run a rectangular bounding box challenge (which we'll do), the workflow and the submissions will be backward incompatible. This tupling is a very small prize to pay for not worrying about backward incompatibility.

**Comment by kegl** [comment]
Anyway, I'm fine with the flat tuple, it just hurts my structural mind :). You can always assume that the first n-1 parameters describe the bounding region, and the last one is the confidence.

**Comment by jorisvandenbossche** [comment]
Something like this can be added later to the scores types that make it possible to extend it and that is backwards compatible:

```patch
diff --git a/workflow/scores/precision_recall.py b/workflow/scores/precision_recall.py
index 23490e4..d0649cb 100644
--- a/workflow/scores/precision_recall.py
+++ b/workflow/scores/precision_recall.py
@@ -220,18 +220,23 @@ class Precision(BaseScoreType):
     minimum = 0.0
     maximum = 1.0
 
-    def __init__(self, name='precision', precision=2, conf_threshold=0.5):
+    def __init__(self, name='precision', precision=2, conf_threshold=0.5,
+                 bbox_type='circular'):
         self.name = name
         self.precision = precision
         self.conf_threshold = conf_threshold
+        self.bbox_type = bbox_type
 
     def __call__(self, y_true, y_pred, conf_threshold=None):
         if conf_threshold is None:
             conf_threshold = self.conf_threshold
-        y_pred_temp = [
-            [(x, y, r) for (x, y, r, p) in y_pred_patch if p > conf_threshold]
-            for y_pred_patch in y_pred]
-        return precision(y_true, y_pred_temp)
+        if self.bbox_type =='circular':
+            y_pred_temp = [
+                [(x, y, r) for (x, y, r, p) in y_pred_patch if p > conf_threshold]
+                for y_pred_patch in y_pred]
+        else:
+            ...
+        return precision(y_true, y_pred_temp, bbox_type=self.bbox_type)
 ```

BTW, the current `precision`, `ospa` (etc) implementation also assumes the specific type of bbox, so will have to be updated to deal with rectangular bboxes anyhow. With the above you can more transparently pass the spec down to the actual score implementation.

> Anyway, I'm fine with the flat tuple, it just hurts my structural mind :). You can always assume that the first n-1 parameters describe the bounding region, and the last one is the confidence.

Yep, although I think typically they put the score first (what I have seen in other challenges / model implementations), and then "all the rest" is the bbox specification

**Comment by kegl** [comment]
I'm ok with p first specs.

**Comment by jorisvandenbossche** [comment]
Just thinking of something else, in a 'typical' object detection workfow, you might also have multiple classes to detect. So this even further complicates how the tuple would look like (I suppose you then have multiple confidence values for the different classes, or the combination of class id (the one with the highest confidence) and the confidence. 
Not sure how that would work with either of the options.

**Comment by kegl** [comment]
The previous commit is important. The scp normalization was wrong even without the minipatch if we have bounding regions not contained completely in the image. Now we'll normalize not by the sum of the _numbers_ of real and detected craters, but the _integrals_ within the minipatches. It's important since the max L1 distance is not 2 but the sum of the partial overlaps between the bounding regions and the minipatch. This made a substantial difference: scp was about 0.7 with the old normalization, now we are back to the correct 0.9976 (with my terrible `full` detector).

**Comment by jorisvandenbossche** [comment]
It would fail with the ``((x, y, r), p)`` structure, unless you make `p` a tuple itself. Then you get something like ``((x, y, r), (class_id, conf))``. Personally I don't really like those nested tuples :-) 
But still you would need to update `DetectionBaseScoreType` somehow to update the way the conf selection is made (and also just the way scores are calculated, as you probably want separate scores for different classes). 

But also with my suggestion of keeping it flat with a possible future `bbox_type` arg, you have exactly the same problems that the current design does not really fit that, so maybe we should leave that for now :-)

**Comment by jorisvandenbossche** [comment]
BTW, I like the `DetectionBaseScoreType`, it really limits the number of times this selection of y_pred based on the conf is done (and also much easier to change it either way to (x, y, r, c) or ((x, y, r), c)

**Comment by kegl** [comment]
Tuples of tuples would make even more sense in that case (one tuple for class posteriors, one for bounding region).

Anyway, if we are agile (solving the problem we have now), then let's be agile and think about it later :)

**Comment by kegl** [comment]
So, are we OK with (p, x, y, r)?

**Comment by kegl** [comment]
@jorisvandenbossche this one is crashing now because `curves.py` also needs to be rewritten. Is there any way we can share code with `DetectionBaseScoreType`?

**Comment by jorisvandenbossche** [comment]
So added a commit that fixes the curves for now.

>  Is there any way we can share code with DetectionBaseScoreType?

I moved the conf filtering to a single function that is reused inside `curves.py`, do the duplication of code is only in a single place now. 
But, in principle I can reuse even more of the code by doing the following:

change 

```
for conf_threshold in conf_thresholds:
        y_pred_above_confidence = _filter_y_pred(y_pred, conf_threshold)
        precision(y_true, y_pred_above_confidence)
        recall(y_true, y_pred_above_confidence)
```

to
```
precision = Precision()

for conf_threshold in conf_thresholds:
        precision(y_true, y_pred, conf_threshold)
        recall(y_true, y_pred, conf_threshold)
```

if we pass through conf_threshold inside the `detection_score` of the DetectionBaseScoreType. 

It gives a little bit less code, but the current code prevents having to redo the conf filtering for both precision and recall.

**Comment by aboucaud** [comment]
Ok need to go for now, @jorisvandenbossche could you check what I just commited, plz ?

Pay especially attention to x,y => row,col / col,row I got confused again. 😄

**Comment by kegl** [comment]
ospa is crashing here

**Comment by kegl** [comment]
OK, it's fixed. But I also thought we would make the OSPA asymmetric. Right now, if a true crater is in the minipatch and it's detected but the predicted crater is outside, you penalize it. I wouldn't. Also if the true crater is outside, it is detected, and the detected crater is in the minipatch, you penalize it. I wouldn't.

**Comment by kegl** [comment]
The same for precision / recall. We agreed that for recall we're counting the number of real craters within the minipatch, matched by a detected crater anywhere in the patch, and for precision, we are counting the number of detected craters in the minipatch, matching real craters anywhere in the patch.

**Comment by jorisvandenbossche** [comment]
This branch PR was getting a bit large, so I merged it apart from the last commits added by @kegl, and made for those last commits a new PR: https://github.com/ramp-kits/mars_craters/pull/27

**Pull request closed without merging**